cpp/src/NVStrings_array.cu

```diff
@@ -20,87 +20,94 @@
+        [d_strings, d_buffer, d_offsets, d_results] __device__(unsigned int idx){
+            custring_view* dstr = d_strings[idx];
+            if( !dstr )
+                return;
+            char* buffer = d_buffer + d_offsets[idx];
+            d_results[idx] = custring_view::create_from(buffer,*dstr);
+        });
+    //
+}
 
 // create a new instance containing only the strings at the specified positions
 // position values can be in any order and can even be repeated
-NVStrings* NVStrings::gather( const int* pos, unsigned int elems, bool bdevmem )
+NVStrings* NVStrings::gather( const int* pos, unsigned int elements, bool bdevmem )
 {
     unsigned int count = size();
-    if( count==0 || elems==0 || pos==0 )
+    if( count==0 || elements==0 || pos==0 )
         return new NVStrings(0);
 
     auto execpol = rmm::exec_policy(0);
     const int* d_pos = pos;
     if( !bdevmem )
     {   // copy indexes to device memory
-        d_pos = const_cast<const int*>(device_alloc<int>(elems,0));
-        CUDA_TRY(cudaMemcpyAsync((void*)d_pos,pos,elems*sizeof(int),cudaMemcpyHostToDevice))
+        d_pos = const_cast<const int*>(device_alloc<int>(elements,0));
+        CUDA_TRY(cudaMemcpyAsync((void*)d_pos,pos,elements*sizeof(int),cudaMemcpyHostToDevice))
     }
-    // get individual sizes
-    rmm::device_vector<long> sizes(elems,0);
-    long* d_sizes = sizes.data().get();
-    custring_view** d_strings = pImpl->getStringsPtr();
-    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), elems,
-        [d_strings, d_pos, count, d_sizes] __device__(unsigned int idx){
+    // create working memory
+    rmm::device_vector<custring_view*> results(elements,nullptr);
+    auto d_results = results.data().get();
+    rmm::device_vector<bool> flags(elements,false);
+    auto d_flags = flags.data().get();
+    custring_view_array d_strings = pImpl->getStringsPtr();
+    // do the gather
+    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), elements,
+        [d_strings, d_pos, count, d_results, d_flags] __device__(unsigned int idx){
             int pos = d_pos[idx];
             if( (pos < 0) || (pos >= count) )
-            {
-                d_sizes[idx] = -1;
-                return;
-            }
-            custring_view* dstr = d_strings[pos];
-            if( dstr )
-                d_sizes[idx] = ALIGN_SIZE(dstr->alloc_size());
+                d_flags[idx] = true;
+            else
+                d_results[idx] = d_strings[pos];
         });
-    // check for any out-of-range values
-    long* first = thrust::min_element(execpol->on(0),d_sizes,d_sizes+elems);
-    long hfirst = 0;
-    CUDA_TRY(cudaMemcpyAsync(&hfirst,first,sizeof(long),cudaMemcpyDeviceToHost))
-    if( hfirst < 0 )
+    // check for invalid position values
+    if( thrust::count(execpol->on(0), flags.begin(), flags.end(), true) )
```

Comment:
I think this is unnecessary. If a user passes in an out of bounds scatter map, then that's their problem. We don't do any bounds checking for libcudf scatter.

Reply:
Done

Reply:
Beautiful